File: los-platform/backend/auth-service/src/controllers/auth.controller.ts
import {
  Controller,
  Post,
  Get,
  Body,
  Req,
  HttpCode,
  HttpStatus,
  Logger,
  UseGuards,
  Query,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse, ApiHeaders } from '@nestjs/swagger';
import { Request } from 'express';
import { AuthService } from '../services/auth.service';
import {
  SendOtpDto,
  VerifyOtpDto,
  RefreshTokenDto,
  SendOtpResponseDto,
  LoginResponseDto,
  RevokeTokenDto,
  RevokeAllSessionsDto,
  LdapLoginDto,
} from '../dto';
import { Public, createError } from '@los/common';

@ApiTags('Authentication')
@Controller('auth')
export class AuthController {
  private readonly logger = new Logger(AuthController.name);

  constructor(private readonly authService: AuthService) {}

  @Public()
  @Post('otp/send')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Send OTP to mobile number' })
  @ApiResponse({ status: 200, description: 'OTP sent successfully', type: SendOtpResponseDto })
  @ApiResponse({ status: 429, description: 'Rate limit exceeded' })
  async sendOtp(
    @Body() dto: SendOtpDto,
    @Req() req: Request,
  ): Promise<SendOtpResponseDto> {
    const ipAddress = this.getClientIp(req);
    const userAgent = req.headers['user-agent'] || '';

    this.logger.log(`OTP send request for ${dto.mobile} from ${ipAddress}`);

    return this.authService.sendOtp(dto, ipAddress, userAgent);
  }

  @Public()
  @Post('otp/verify')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Verify OTP and receive JWT tokens' })
  @ApiResponse({ status: 200, description: 'Authentication successful', type: LoginResponseDto })
  @ApiResponse({ status: 401, description: 'Invalid or expired OTP' })
  async verifyOtp(
    @Body() dto: VerifyOtpDto,
    @Req() req: Request,
  ): Promise<LoginResponseDto> {
    const ipAddress = this.getClientIp(req);

    this.logger.log(`OTP verify request for session ${dto.sessionId}`);

    return this.authService.verifyOtp(dto, ipAddress);
  }

  @Public()
  @Post('ldap/login')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Bank staff LDAP/AD login (branch manager, loan officer, credit analyst)' })
  @ApiResponse({ status: 200, description: 'Authentication successful', type: LoginResponseDto })
  @ApiResponse({ status: 401, description: 'Invalid credentials' })
  async ldapLogin(@Body() dto: LdapLoginDto): Promise<LoginResponseDto> {
    this.logger.log(`LDAP login attempt for user: ${dto.username}`);
    return this.authService.ldapLogin(dto.username, dto.password, dto.deviceFingerprint);
  }

  @Public()
  @Post('token/refresh')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Refresh access token using refresh token' })
  @ApiResponse({ status: 200, description: 'Tokens refreshed', type: LoginResponseDto })
  @ApiResponse({ status: 401, description: 'Invalid or expired refresh token' })
  async refreshToken(@Body() dto: RefreshTokenDto): Promise<LoginResponseDto> {
    return this.authService.refreshToken(dto);
  }

  @Post('logout')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: 'Logout and revoke all sessions' })
  @ApiHeaders([
    { name: 'Authorization', description: 'Bearer token', required: true },
  ])
  @ApiResponse({ status: 204, description: 'Logged out successfully' })
  async logout(@Req() req: Request): Promise<void> {
    const user = req.user as any;
    const sessionId = user?.sessionId;

    await this.authService.logout(user?.sub, sessionId);
  }

  @Post('revoke')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Revoke a specific token (access token or JTI)' })
  @ApiResponse({ status: 200, description: 'Token revoked' })
  async revokeToken(@Body() dto: RevokeTokenDto): Promise<{ success: boolean }> {
    await this.authService.revokeToken(dto.tokenOrJti, dto.reason || 'User requested revocation');
    return { success: true };
  }

  @Post('sessions/revoke-all')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Revoke all sessions for a user (admin)' })
  @ApiResponse({ status: 200, description: 'Sessions revoked', schema: { properties: { revokedCount: { type: 'number' } } } })
  async revokeAllSessions(@Body() dto: RevokeAllSessionsDto): Promise<{ revokedCount: number }> {
    const count = await this.authService.revokeAllUserSessions(dto.userId, dto.exceptSessionId);
    return { revokedCount: count };
  }

  private getClientIp(req: Request): string {
    return (
      (req.headers['x-forwarded-for'] as string)?.split(',')[0]?.trim() ||
      (req.headers['x-real-ip'] as string) ||
      req.socket.remoteAddress ||
      'unknown'
    );
  }
}

File: los-platform/backend-java/src/main/java/com/los/auth/controller/AuthController.java
package com.los.auth.controller;

import com.los.auth.dto.*;
import com.los.auth.service.AuthService;
import com.los.auth.service.LdapAuthService;
import com.los.common.dto.ApiResponse;
import com.los.common.security.JwtTokenProvider;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@Slf4j
@RestController
@RequestMapping("/api/auth")
@RequiredArgsConstructor
@Tag(name = "Authentication", description = "Authentication and authorization endpoints")
public class AuthController {

    private final AuthService authService;
    private final LdapAuthService ldapAuthService;
    private final JwtTokenProvider jwtTokenProvider;

    /**
     * Send OTP to mobile number
     */
    @PostMapping("/otp/send")
    @Operation(summary = "Send OTP", description = "Send OTP to registered mobile number for login or verification")
    @ApiResponses(value = {
        @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "200", description = "OTP sent successfully"),
        @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "400", description = "Invalid request"),
        @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "429", description = "Too many requests")
    })
    public ResponseEntity<ApiResponse<java.util.Map<String, String>>> sendOtp(
            @Valid @RequestBody SendOtpDto dto,
            HttpServletRequest request) {
        log.info("OTP send request for mobile");

        dto.setIpAddress(getClientIp(request));
        dto.setUserAgent(request.getHeader("User-Agent"));

        ApiResponse<java.util.Map<String, String>> response = authService.sendOtp(dto);
        return ResponseEntity.ok(response);
    }

    /**
     * Verify OTP and login
     */
    @PostMapping("/otp/verify")
    @Operation(summary = "Verify OTP and Login", description = "Verify OTP and get access token for authenticated user")
    public ResponseEntity<ApiResponse<LoginResponseDto>> verifyOtp(
            @Valid @RequestBody VerifyOtpDto dto,
            HttpServletRequest request) {
        log.info("OTP verification request");

        dto.setIpAddress(getClientIp(request));
        dto.setUserAgent(request.getHeader("User-Agent"));

        ApiResponse<LoginResponseDto> response = authService.verifyOtpAndLogin(dto);
        return ResponseEntity.ok(response);
    }

    /**
     * LDAP login for employees
     */
    @PostMapping("/ldap/login")
    @Operation(summary = "LDAP Login", description = "Login using LDAP credentials for internal employees")
    public ResponseEntity<ApiResponse<LoginResponseDto>> ldapLogin(
            @Valid @RequestBody LdapLoginDto dto,
            HttpServletRequest request) {
        log.info("LDAP login request");

        dto.setIpAddress(getClientIp(request));
        dto.setUserAgent(request.getHeader("User-Agent"));

        ApiResponse<LoginResponseDto> response = ldapAuthService.ldapLogin(dto);
        return ResponseEntity.ok(response);
    }

    /**
     * Refresh access token
     */
    @PostMapping("/token/refresh")
    @Operation(summary = "Refresh Token", description = "Get new access token using valid refresh token")
    public ResponseEntity<ApiResponse<LoginResponseDto>> refreshToken(
            @Valid @RequestBody RefreshTokenDto dto,
            HttpServletRequest request) {
        log.info("Token refresh request");

        dto.setDeviceFingerprint(request.getHeader("X-Device-Fingerprint"));

        ApiResponse<LoginResponseDto> response = authService.refreshToken(dto);
        return ResponseEntity.ok(response);
    }

    /**
     * Logout user
     */
    @PostMapping("/logout")
    @Operation(summary = "Logout", description = "Logout and revoke access token")
    public ResponseEntity<ApiResponse<Void>> logout(
            @RequestHeader("Authorization") String authHeader) {
        log.info("Logout request");

        String token = authHeader.replace("Bearer ", "");
        ApiResponse<Void> response = authService.logout(token, "User logout");
        return ResponseEntity.ok(response);
    }

    /**
     * Revoke specific token
     */
    @PostMapping("/revoke")
    @Operation(summary = "Revoke Token", description = "Revoke a specific token or session")
    public ResponseEntity<ApiResponse<Void>> revokeToken(
            @Valid @RequestBody RevokeTokenDto dto) {
        log.info("Token revoke request");
        ApiResponse<Void> response = authService.revokeToken(dto);
        return ResponseEntity.ok(response);
    }

    /**
     * Revoke all user sessions
     */
    @PostMapping("/sessions/revoke-all")
    @Operation(summary = "Revoke All Sessions", description = "Revoke all active sessions for a user")
    public ResponseEntity<ApiResponse<Void>> revokeAllSessions(
            @Valid @RequestBody RevokeAllSessionsDto dto) {
        log.info("Revoke all sessions request for user: {}", dto.getUserId());
        ApiResponse<Void> response = authService.revokeAllSessions(dto);
        return ResponseEntity.ok(response);
    }

    /**
     * Get JWKS for token validation
     */
    @GetMapping("/.well-known/jwks.json")
    @Operation(summary = "Get JWKS", description = "Get JSON Web Key Set for token validation")
    public ResponseEntity<java.util.Map<String, Object>> getJwks() {
        log.debug("JWKS endpoint called");
        return ResponseEntity.ok(java.util.Map.of(
            "keys", java.util.List.of(jwtTokenProvider.getPublicJwk())
        ));
    }

    /**
     * Get current user profile
     */
    @GetMapping("/profile")
    @Operation(summary = "Get Profile", description = "Get current authenticated user's profile information")
    public ResponseEntity<ApiResponse<AuthenticatedUserDto>> getProfile() {
        Object principal = org.springframework.security.core.context.SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        if (!(principal instanceof com.los.common.security.AuthenticatedUser authenticatedUser)) {
            throw new com.los.common.exception.LosException("AUTH_008", "Not authenticated", 401, false);
        }
        
        ApiResponse<AuthenticatedUserDto> response = authService.getProfile(authenticatedUser.getId());
        return ResponseEntity.ok(response);
    }

    private String getClientIp(HttpServletRequest request) {
        String clientIp = request.getHeader("X-Forwarded-For");
        if (clientIp == null || clientIp.isEmpty() || "unknown".equalsIgnoreCase(clientIp)) {
            clientIp = request.getRemoteAddr();
        }
        return clientIp;
    }
}
